### ISSUE TITLE
prxs fails to detect module-specific configure/Makefile, leading to unexpected module load errors

### ISSUE THREAD

**Issue by hpreusse** [proftpd/proftpd#1079]
### What I Did

I've build the proftpd-mod-proxy module for Debian and now I try to load the module. I put the statement `#LoadModule mod_proxy.c` into the `/etc/proftpd/modules.conf`, but the module fails to load.

### What I Expected/Wanted

I would have expected that the module get's loaded and the features are available.

### ProFTPD Version and Configuration

Here is the error message I get in the logs when the module fails to load:

```
root@amd64-sid:/etc/proftpd# service proftpd check-config
2020-08-01 20:43:20,950 amd64-sid proftpd[5836]: mod_dso/0.5: module 'mod_shaper.c' already loaded
2020-08-01 20:43:20,950 amd64-sid proftpd[5836]: mod_dso/0.5: unable to load 'mod_proxy.c'; check to see
 if '/usr/lib/proftpd/mod_proxy.la' exists
2020-08-01 20:43:20,951 amd64-sid proftpd[5836]: fatal: LoadModule: error loading module 'mod_proxy.c':
No such file or directory on line 110 of '/etc/proftpd/modules.conf'
2020-08-01 20:43:20,951 amd64-sid proftpd[5836]: warning: unable to include '/etc/proftpd/modules.conf':
 Operation not permitted
ProFTPD configuration OK
```
Copying the file /usr/lib/proftpd/mod_proxy.la' into `/usr/lib/proftpd/` just changes the error message:

```
root@amd64-sid:/usr/lib/proftpd# service proftpd check-config
2020-08-01 20:45:44,760 amd64-sid proftpd[5951]: mod_dso/0.5: module 'mod_shaper.c' already loaded
2020-08-01 20:45:44,761 amd64-sid proftpd[5951]: fatal: LoadModule: error loading module 'mod_proxy.c':
Invalid argument on line 110 of '/etc/proftpd/modules.conf'
2020-08-01 20:45:44,761 amd64-sid proftpd[5951]: warning: unable to include '/etc/proftpd/modules.conf':
 Operation not permitted
ProFTPD configuration OK
```
A similar bug report was opened @ Debian 3 days after we uploaded 1.3.6c: [#951916](https://bugs.debian.org/951916). Hence I suspect that something broke in 1.3.6c.

Here is the `proftpd -V`, the `proftpd.conf` and the `modules.conf` are attached. No, there are no config parameters for that module. I'd have expected that a simple load should have worked without them. Many thanks!
```
root@amd64-sid:~# proftpd -V
Compile-time Settings:
  Version: 1.3.7a (maint)
  Platform: LINUX [Linux 5.7.0-2-amd64 x86_64]
  Built: Thu Jul 23 2020 05:53:17 UTC
  Built With:
    configure  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-option-checking' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--with-includes=/usr/include/postgresql:/usr/include/mariadb:/usr/include/mariadb/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp:mod_digest:mod_ident:mod_log_forensic:mod_qos:mod_statcache:mod_tls_fscache:mod_tls_shmcache' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-ynkZ3Q/proftpd-dfsg-1.3.7a=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-ynkZ3Q/proftpd-dfsg-1.3.7a=. -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-ynkZ3Q/proftpd-dfsg-1.3.7a=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib -Wl,-z,relro -rdynamic  -L/usr/lib/x86_64-linux-gnu/ -L/usr/lib/x86_64-linux-gnu
  LIBS: -lacl  -lpcreposix -lpcre -lssl -lcrypto -lsodium -lcap  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -ldl -lhiredis -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 1.1.1g  21 Apr 2020)
    + PCRE support
    + POSIX ACL support
    + Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```
[proftpd.conf.txt](https://github.com/proftpd/proftpd/files/5031425/proftpd.conf.txt)
[modules.conf.txt](https://github.com/proftpd/proftpd/files/5031457/modules.conf.txt)

**Comment by Castaglia** [comment]
You cannot just copy the `.la` file -- there's more to installing 3rd-party modules than that.  Did you follow all of the `mod_proxy` [installation instructions](https://htmlpreview.github.io/?https://github.com/Castaglia/proftpd-mod_proxy/blob/master/mod_proxy.html#Installation)?

**Comment by hpreusse** [comment]
In Debian we use prxs to build additional modules like this:

`./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var`

`DESTDIR=$(CURDIR)/debian/proftpd-mod-proxy prxs -c -I include mod_proxy.c`
```
DESTDIR=$(CURDIR)/debian/proftpd-mod-proxy prxs -i -I include -c mod_proxy.c
  rm -f $(CURDIR)/debian/proftpd-mod-proxy/usr/lib/proftpd/*.la \
           $(CURDIR)/debian/proftpd-mod-proxy/usr/lib/proftpd/*.a
```
During build we have the package [proftpd-dev](https://packages.debian.org/sid/proftpd-dev) installed, which provides static libs and header files.

**Comment by Castaglia** [comment]
I am not sure that `prxs` works properly for multi-file modules like `mod_proxy.c`; I have not explicitly tested/used/verified that it works for such modules.  Thus I cannot, at this time, recommend using `prxs` for such modules; see [Bug #3308](http://bugs.proftpd.org/show_bug.cgi?id=3308).

**Comment by hpreusse** [comment]
Many thanks for explanation! I could compile the module w/o prxs and the whole build tree. The module was loadable.

2020-08-13 00:17:12,625 debian-amd64-sid proftpd[63655]: mod_dso/0.5: loading 'mod_vroot.c'
2020-08-13 00:17:12,625 debian-amd64-sid proftpd[63655]: mod_dso/0.5: loaded module 'mod_vroot' (from '/usr/lib/proftpd/mod_vroot.so', last modified on Thu Aug 13 00:00:59 2020)

**Comment by Castaglia** [comment]
@hpreusse When I use `prxs` to try to build _e.g._ `mod_proxy`, following your instructions, I see this:
```
$ prxs -c -I /usr/local/proftpd/include mod_proxy.c
Cannot compile mod_proxy using prxs; use configure instead
$ echo $?
```
Perhaps I am not reproducing the same steps/commands that the Debian build system is doing for these modules?  I ask because I'd like to address [Bug #3308](http://bugs.proftpd.org/show_bug.cgi?id=3308), one way or another.

**Comment by hpreusse** [comment]
Still same behaviour in old build system using `prxs`:

```
hille@sid:~/devel/proftp/aaa_git/proftpd-mod-proxy $ ./configure --with-modules=mod_proxy --enable-dso
<snip>
configure: creating ./config.status
config.status: creating t/Makefile
config.status: creating Makefile
config.status: creating mod_proxy.h
hille@sid:~/devel/proftp/aaa_git/proftpd-mod-proxy $ prxs -c mod_proxy.c -I /usr/include/proftpd/ -I . -I include
libtool --mode=compile gcc  -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-DhaPyH/proftpd-dfsg-1.3.7a=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -I/usr/include/proftpd/ -I. -Iinclude -c mod_proxy.c
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-DhaPyH/proftpd-dfsg-1.3.7a=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -I/usr/include/proftpd/ -I. -Iinclude -c mod_proxy.c  -fPIC -DPIC -o .libs/mod_proxy.o
libtool: compile:  gcc -DLINUX -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-DhaPyH/proftpd-dfsg-1.3.7a=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -DPR_SHARED_MODULE -I. -I/usr/include/proftpd -I/usr/include/proftpd/ -I. -Iinclude -c mod_proxy.c -o mod_proxy.o >/dev/null 2>&1
libtool --mode=link gcc -o mod_proxy.la -rpath /usr/lib/proftpd  -avoid-version -export-dynamic -module  mod_proxy.lo  -lsqlite3
libtool: link: gcc -shared  -fPIC -DPIC  .libs/mod_proxy.o   -lsqlite3    -Wl,-soname -Wl,mod_proxy.so -o .libs/mod_proxy.so
libtool: link: ar cr .libs/mod_proxy.a  mod_proxy.o
libtool: link: ranlib .libs/mod_proxy.a
libtool: link: ( cd ".libs" && rm -f "mod_proxy.la" && ln -s "../mod_proxy.la" "mod_proxy.la" )
hille@sid:~/devel/proftp/aaa_git/proftpd-mod-proxy $
```

Assuming that [proftpd-dev](https://packages.debian.org/sid/proftpd-dev) is installed. `./configure` has to run, else the `prxs` complains about missing header files. As one can see, not all source files are compiled, hence one gets an incomplete module.

**Comment by Castaglia** [comment]
@hpreusse Sorry, let me be clearer -- what is the output you see, in your build system, _when `prxs` is used?_

If you see the same message from `prxs`:
```
Cannot compile mod_proxy using prxs; use configure instead
```
in that build system output, then why didn't the build system notice the failure?

**Comment by Castaglia** [comment]
@hpreusse yes, that helped.  Thank you!  Would you be able to try out [this PR](https://github.com/proftpd/proftpd/pull/1094), for the above steps, to help verify that with that change applied, `prxs` properly detects the presence of the `configure` script for the module?

**Comment by hpreusse** [comment]
With the patch attached it looks better. Now the build fails, if there is a configure script present. Please go ahead and merge the request. Many thanks for help!

**Comment by Castaglia** [comment]
Fixes backported to 1.3.7 branch, and merged to master.  Thanks!
